node/network/approval-distribution/src/lib.rs

```diff
@@ -1674,83 +1811,78 @@ async fn adjust_required_routing_and_propagate(
 	// Send messages in accumulated packets, assignments preceding approvals.
 
 	for (peer, assignments_packet) in peer_assignments {
-		ctx.send_message(NetworkBridgeMessage::SendValidationMessage(
+		ctx.send_message(NetworkBridge(NetworkBridgeMessage::SendValidationMessage(
 			vec![peer],
 			protocol_v1::ValidationProtocol::ApprovalDistribution(
 				protocol_v1::ApprovalDistributionMessage::Assignments(assignments_packet),
 			),
-		))
+		)))
 		.await;
 	}
 
 	for (peer, approvals_packet) in peer_approvals {
-		ctx.send_message(NetworkBridgeMessage::SendValidationMessage(
+		ctx.send_message(NetworkBridge(NetworkBridgeMessage::SendValidationMessage(
 			vec![peer],
 			protocol_v1::ValidationProtocol::ApprovalDistribution(
 				protocol_v1::ApprovalDistributionMessage::Approvals(approvals_packet),
 			),
-		))
+		)))
 		.await;
 	}
 }
 
 /// Modify the reputation of a peer based on its behavior.
-async fn modify_reputation(
-	ctx: &mut (impl SubsystemContext<Message = ApprovalDistributionMessage>
-	          + overseer::SubsystemContext<Message = ApprovalDistributionMessage>),
-	peer_id: PeerId,
-	rep: Rep,
-) {
+async fn modify_reputation(ctx: &mut impl SubsystemSender, peer_id: PeerId, rep: Rep) {
 	gum::trace!(
 		target: LOG_TARGET,
 		reputation = ?rep,
 		?peer_id,
 		"Reputation change for peer",
 	);
 
-	ctx.send_message(NetworkBridgeMessage::ReportPeer(peer_id, rep)).await;
+	ctx.send_message(NetworkBridge(NetworkBridgeMessage::ReportPeer(peer_id, rep)))
+		.await;
 }
 
-impl ApprovalDistribution {
+impl ApprovalDistributionWorker {
 	/// Create a new instance of the [`ApprovalDistribution`] subsystem.
-	pub fn new(metrics: Metrics) -> Self {
-		Self { metrics }
+	pub fn new(
+		metrics: Metrics,
+		proxy_receiver: futures::channel::mpsc::Receiver<FromOverseer<ApprovalDistributionMessage>>,
+		worker_index: usize,
+	) -> Self {
+		Self { metrics, proxy_receiver, worker_index }
 	}
 
-	async fn run<Context>(self, ctx: Context)
-	where
-		Context: SubsystemContext<Message = ApprovalDistributionMessage>,
-		Context: overseer::SubsystemContext<Message = ApprovalDistributionMessage>,
-	{
-		let mut state = State::default();
-
+	#[tokio::main]
```

Comment:
what's the purpose of tokio::main here considering we're manually spawning the threads? (I understand that both things are not what we want to do in production and are only used for this test, but curious why ::main is needed at all)

Reply:
It is not needed, it is just a leftover. I'll switch it to tasks. I hoped that I can hold the lock while awaiting if I use threads...